### DISCUSSION TITLE
Custom UI modding via XML

### DISCUSSION THREAD

**Discussion by neirenoir** [FakeFishGames/Barotrauma#12997] [Ideas and suggestions]
Right now, whenever some modder wants to tweak or add a new UI for a component, they need to crack open the C# source code and re-write the entire thing to be used with LuaCsForBarotrauma. This is cumbersome for both the developers of the mod, and the users, who have to go through some extra hoops to enable C# mods (and for good reason!).

While some UI are too complex to be written in just XML, like the sonar section of the navigation console, there is no particular reason to have simpler windows be hardcoded, like the Health UI or even the Reactor one.

While it would be kind of a pain in the ass to move most current UI to a new XML-based defined one, it would make sense. XML has been traditionally used to describe UI, and it could be argued it is the number 1 language for GUI description right now if we count HTML as a subset of XML. Given almost the entirety of the game is written in XML already, it would make sense to move the GUI definitions of many components to XML files, while maybe retaining some of the more complex ones (ie. the sonar, the reactor needle indicators) as prebuilt components in C# referenced inside of the XML UI file as a special tag.

This is probably a massive endeavor, but it would allow mods to set up custom interfaces and, for example, help with making the Captain's job more interesting by adding more controls to their main console, or help mod developers make more complex Reactor mods that could allow for more in-depth reactor control for engineers (think IndustrialCraft2's reactor matrix configuration system)

**Comment by Alkanty** [comment]
i 100 % agree this. 

adding more modding tool including making UI more moddable could make life 100 % easier for modder while bringing back more people to barotrauma . a lot of game continue to be played and bought because of community modding (think of Empire at war, Skyrim, stalker, Starcraft 2,Arma where you have total conversion mod). The UI change enable Gameplay change that are a pain to make at the moment. 

i finished the campaign with friends but keep playing because of mods (in my 120 h of game 90h are with mods) 
-replayability comes from good mod that comes from good tools ^^
-Could be easier to make DLC and Expansion with some enchanced UI tools
- Baro has great tools and XML is easy to understand. UI and some elements in the game are still hardcoded tho.

its an opinion !

**Comment by iridescent-lightning** [comment]
The sonar panel has a function that adjusts its elements' positions if a new switch (mineral switch) is added. The size of these elements are relative rectransform, which means that if you mess with the panel, you get the wrong level size and switch size. Then you have to add new control to auto adjust pre-exist elements if new elements are added. This cannot be done in xml.
